### PR TITLE
[WIP] Send and Sync for Fence and Mapping

### PR DESCRIPTION
**Pull request by Bastacyclop**
Here's what I made for core and gl, let me know if you don't like something.
@kvark

### CODE REVIEW

File: src/backend/gl/src/factory.rs [hunk 1]
```diff
@@ -210,46 +210,22 @@ impl Factory {
...
 
-#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
 pub enum MappingKind {
     Persistent,
     Temporary,
 }
 
-#[derive(Clone)]
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
```
Comment (Bastacyclop) [inline src/backend/gl/src/factory.rs]: `Copy` was probably too much

File: src/backend/gl/src/lib.rs [hunk 2]
```diff
@@ -921,6 +875,55 @@ impl c::Device for Device {
         fence_opt.unwrap_or_else(|| self.place_fence())
     }
 
+    fn wait_fence(&mut self, fence: &handle::Fence<Self::Resources>) {
+        wait_fence(self.frame_handles.ref_fence(&fence), &self.share.context);
+    }
+
+    fn read_mapping<'a, 'b, M, T>(&'a mut self, m: &'b mut M)
```
Comment (kvark) [inline src/backend/gl/src/lib.rs]: Please move the mapping related methods to the `Factory`. The device should be responsible for commands submission, not memory management.

File: src/backend/gl/src/lib.rs [hunk 3]
```diff
@@ -940,14 +943,26 @@ impl c::Device for Device {
             |_, _| {}, //RTV
             |_, _| {}, //DSV
             |gl, v| unsafe { if v.object != 0 { gl.DeleteSamplers(1, &v.object) }},
-            |gl, fence| unsafe { gl.DeleteSync(fence.raw.0) },
-            |_, raw_mapping| {
+            |gl, fence| unsafe { gl.DeleteSync(fence.0) },
+            |&mut gl, raw_mapping| {
```
Comment (kvark) [inline src/backend/gl/src/lib.rs]: is `&mut` needed here?

File: examples/instancing/main.rs [hunk 4]
```diff
@@ -150,20 +141,10 @@ impl<R: gfx::Resources> gfx_app::Application<R> for App<R> {
...
             slice: slice,
-            mapping: instance_mapping,
-            counter: 0,
         }
     }
 
     fn render<C: gfx::CommandBuffer<R>>(&mut self, encoder: &mut gfx::Encoder<R, C>) {
-        if self.counter == 120 { // TODO: use frame delta time ?
-            let mut instances = self.mapping.read_write();
-            update_instances(&mut instances);
```
Comment (kvark) [inline examples/instancing/main.rs]: so now we are never updating instances?

File: examples/shadow/main.rs [hunk 5]
```diff
@@ -527,8 +527,6 @@ impl<R, C> gfx_app::ApplicationBase<R, C> for App<R, C> where
 
         // fill up shadow map for each light
         if self.is_parallel {
-            unimplemented!() // FIXME: add mapping Send bound
-            /*
```
Comment (kvark) [inline examples/shadow/main.rs]: yay!

File: src/backend/dx11/src/factory.rs [hunk 6]
```diff
@@ -27,11 +27,14 @@ use {CommandList, DeferredContext};
 use native;
 
 
-#[derive(Copy, Clone, Debug)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
```
Comment (kvark) [inline src/backend/dx11/src/factory.rs]: let's not have it `Copy`, as you suggested

**Comment by kvark** [comment]
@homu r+

**Pull request merged**